### PR TITLE
Dwell-times from Gaussian Mixture Models

### PR DESCRIPTION
**Pull request by rpauszek**
**Why this PR?**
We have two main functionalities currently in the `population` module: Gaussian mixture models and Dwell-time analysis. This feature allows for the results from the first to be used as input to the second. Specifically, channel data can be fit with the GMM and a labeled state-path extracted. This state-path is converted into lists of dwell-times for each state and these can then be fit with the exponential distribution.

While I was working on the tutorial update I noticed the need for this feature. So there are no docs updates here; a second PR with those changes will follow shortly

It might also be useful to have some classmethod on `DwelltimeModel` to read in state-paths directly

### CODE REVIEW

File: lumicks/pylake/population/dwelltime.py [hunk 1]
```diff
@@ -480,3 +480,55 @@ def _exponential_mle_optimize(
...
+    dwelltimes, multiply the counts by the time step between points.
+
+    Parameters
+    ----------
+    statepath : numpy.typing.ArrayLike
+        Time-ordered array of state labels
+    exclude_ambiguous_dwells : bool
+            Determines whether to exclude dwelltimes which are not exactly determined. If `True`, tracks which
+            start in the first frame or end in the last frame of the kymograph are not used in the analysis,
+            since the exact start/stop times of the binding event are not definitively known.
```
Comment (JoepVanlier) [inline lumicks/pylake/population/dwelltime.py:498]: Copy/paste? Kymograph doesn't seem relevant here? Also, the indent is not correct.

File: lumicks/pylake/population/tests/test_dwelltimes.py [hunk 2]
```diff
@@ -62,3 +63,39 @@ def test_bootstrap(exponential_data):
     np.testing.assert_allclose(mean, 0.4642469883372174, rtol=1e-5)
     np.testing.assert_allclose(ci, (0.3647038711684928, 0.5979550940729152), rtol=1e-5)
     np.random.seed()
+
+
+def test_dwellcounts_from_statepath():
```
Comment (JoepVanlier) [inline lumicks/pylake/population/tests/test_dwelltimes.py:68]: Include a test for an empty state path?

File: lumicks/pylake/population/mixture.py [hunk 3]
```diff
@@ -121,6 +122,26 @@ def label(self, trace):
...
+        Parameters
+        ----------
+        statepath : numpy.typing.ArrayLike
+            Time-ordered array of state labels. This can be obtained using the `GaussianMixtureModel.label()` method.
+        dt_seconds :
+            Time between data points in seconds.
+        exclude_ambiguous_dwells : bool
+            Determines whether to exclude dwelltimes which are not exactly determined. If `True`, tracks which
+            start in the first frame or end in the last frame of the kymograph are not used in the analysis,
+            since the exact start/stop times of the binding event are not definitively known.
```
Comment (JoepVanlier) [inline lumicks/pylake/population/mixture.py:137]: Copy/paste? Kymograph is mentioned again.

File: lumicks/pylake/population/mixture.py [hunk 4]
```diff
@@ -121,6 +122,26 @@ def label(self, trace):
         output_states = np.argsort(self._map)  # output model state labels in wrapped model order
         return output_states[labels]  # output model labels
 
+    def dwell_times(self, statepath, dt_seconds, exclude_ambiguous_dwells):
```
Comment (JoepVanlier) [inline lumicks/pylake/population/mixture.py:125]: Why is this a method? You don't use anything contained in `self`.

The fact that it doesn't, makes me wonder about the API though. Why not take a slice, then you can label it for the user, extract the time step etc?

Also, would it make sense to make `exclude_ambiguous_dwells` a keyword argument?

File: changelog.md [hunk 5]
```diff
@@ -1,5 +1,10 @@
 # Changelog
 
+## v0.13.3 | tbd
+
+#### New features
+* Added the `lk.GaussianMixtureModel.dwell_times()` method to calculate dwell-times for all states in a time-ordered state trajectory.
```
Comment (JoepVanlier) [inline changelog.md:6]:
```suggestion
#### New features

* Added the `lk.GaussianMixtureModel.dwell_times()` method to calculate dwell-times for all states in a time-ordered state trajectory.
```

**Review by JoepVanlier — CHANGES_REQUESTED** [review]
Nice PR! Especially like the little private method for calculating the dwell times.

Not so sure about the API on the MixtureModel though.

File: lumicks/pylake/population/dwelltime.py [hunk 6]
```diff
@@ -480,3 +480,55 @@ def _exponential_mle_optimize(
...
+
+def _dwellcounts_from_statepath(statepath, exclude_ambiguous_dwells=True):
+    """Calculate the dwell counts and slicing indices for all states in a state path trajectory.
+
+    Note: the counts are the number of frames or time points. To convert to proper
+    dwelltimes, multiply the counts by the time step between points.
+
+    Parameters
+    ----------
+    statepath : numpy.typing.ArrayLike
```
Comment (JoepVanlier) [inline lumicks/pylake/population/dwelltime.py:493]: I think here just using `array_like` is fine? The `numpy.typing` stuff was for the type annotations.

File: lumicks/pylake/population/mixture.py [hunk 7]
```diff
@@ -121,6 +122,26 @@ def label(self, trace):
...
+        """Calculate lists of dwelltimes for each state in a time-ordered statepath array.
+
+        Parameters
+        ----------
+        statepath : numpy.typing.ArrayLike
```
Comment (JoepVanlier) [inline lumicks/pylake/population/mixture.py:130]:
```suggestion
        statepath : array_like
```

File: lumicks/pylake/population/mixture.py (hunk 4 above)
Comment (rpauszek) [reply lumicks/pylake/population/mixture.py]: yeah you're right. I've changed it now and I'll add a section in the docs for how to use this with `DwelltimeModel` in the next PR

File: lumicks/pylake/population/mixture.py [hunk 8]
```diff
@@ -121,6 +122,34 @@ def label(self, trace):
...
+
+        Returns
+        -------
+        dict:
+            Dictionary of all dwell times (in seconds) for each state. Keys are state labels.
+        dict:
+            Dictionary of slicing indices for all dwell ranges for each state. Keys are state labels.
+        """
+        statepath = self.label(trace)
+        dt_seconds = 1 / trace.sample_rate
```
Comment (JoepVanlier) [inline lumicks/pylake/population/mixture.py:145]: nitpick: `1.0 / trace.sample_rate`

**Review by JoepVanlier — APPROVED** [review]
Few minor suggestions but other than that it looks good to me.
